crates/ark/src/analysis/parse_boundaries.rs

```diff
@@ -0,0 +1,375 @@
+                    incomplete_end = Some(current_line + 1);
+                }
+            },
+
+            ParseResult::SyntaxError { .. } => {
+                // Declare error
+                if let None = error_end {
+                    error_end = Some(current_line + 1);
+                }
+            },
+        };
+    }
+
+    // Note that these are necessarily no-ops if we exited the loop after
+    // detected complete expressions since we already recorded the boundaries in
+    // that case.
+    record_incomplete(0, &incomplete_end);
+    record_error(0, &error_end);
+
+    // Merge expressions separated by semicolons
+    let complete = merge_overlapping(complete);
+
+    let boundaries = ParseBoundaries {
+        complete,
+        incomplete,
+        error,
+    };
+
+    // Fill any gaps with one-liner complete expressions
+    let boundaries = fill_gaps(boundaries, n_lines);
+
+    Ok(boundaries)
+}
+
+fn merge_overlapping<T>(ranges: Vec<std::ops::Range<T>>) -> Vec<std::ops::Range<T>>
+where
+    T: PartialOrd + Ord + Copy,
+{
+    let merge = |mut merged: Vec<std::ops::Range<T>>, range: std::ops::Range<T>| {
+        if let Some(last) = merged.last_mut() {
+            // if range.start <= last.end {
+            if last.contains(&range.start) {
+                // Overlap, merge with last range
+                last.end = last.end.max(range.end);
+            } else {
+                // No overlap, push a new range
+                merged.push(range);
+            }
+        } else {
+            // First element, just add it
+            merged.push(range);
+        }
+
+        merged
+    };
+
+    ranges.into_iter().fold(vec![], merge)
+}
+
+// Fill any gaps with one-liner complete expressions. This applies
+// to empty lines, with or without spaces and tabs, and possibly
+// with a trailing comment.
+fn fill_gaps(mut boundaries: ParseBoundaries, n_lines: usize) -> ParseBoundaries {
+    let mut filled = vec![];
+    let ranges = boundaries.complete;
+
+    let mut last_line = 0;
+
+    let range_from = |start| std::ops::Range {
+        start,
+        end: start + 1,
+    };
+
+    // Fill leading whitespace with empty input ranges
+    if let Some(first) = ranges
+        .get(0)
+        .or(boundaries.incomplete.as_ref())
+        .or(boundaries.error.as_ref())
+    {
+        for start in 0..first.start {
+            let range = range_from(start);
+            last_line = range.end;
+            filled.push(range)
+        }
+    }
+
+    // Fill gaps between complete expressions
+    {
+        let iter = ranges.into_iter();
+        for range in iter {
+            // We found a gap, fill ranges for lines in that gap
+            if !range.contains(&last_line) {
+                for start in last_line..range.start {
+                    filled.push(range_from(start))
+                }
+            }
+
+            last_line = range.end;
+            filled.push(range);
+        }
+    }
```

Comment:
```suggestion
    // Fill gaps between complete expressions
    for range in ranges.into_iter() {
        // We found a gap, fill ranges for lines in that gap
        if !range.contains(&last_line) {
            for start in last_line..range.start {
                filled.push(range_from(start))
            }
        }

        last_line = range.end;
        filled.push(range);
    }
```

Doesn't seem like you need the extra scope

Reply:
Right I think I had a `mut` in there at some point that I wanted a scope for, but it's no longer there.